### ISSUE TITLE
[OpenCL] hipStreamPerThread_MultiThread crashes in uninitialization

### ISSUE THREAD

**Issue by pjaaskel** [CHIP-SPV/chip-spv#180]
Started debugging this Intel OpenCL CPU triggered test case:
```
Filters: Unit_hipStreamPerThread_MultiThread
===============================================================================
test cases: 1 | 1 passed
assertions: - none -

Aborted (core dumped)
```
I think this might be somehow related to your latest event refcounting commits @pvelesko.

```C++
603	void CHIPEventOpenCL::increaseRefCount(std::string Reason) {
604	  std::lock_guard<std::mutex> Lock(EventMtx);
605	  auto status = clRetainEvent(this->ClEvent);
606	  if(!UserEvent_)
607	    assert(status == 0);
608	  logDebug("CHIPEventOpenCL::increaseRefCount() {} {} refc {}->{} REASON: {}",
609	           (void *)this, Msg.c_str(), *Refc_, *Refc_ + 1, Reason);
610	  (*Refc_)++;
611	  assert(*Refc_ = getRefCount() - 1);
(gdb) print UserEvent_
$1 = false
...
(gdb) print status
$3 = -58
```
-58 is CL_INVALID_EVENT, so perhaps the event is freed already. It might be a race condition with the async commands on the fly. Something is still in flight while the test has already finished? It's a heisenbug, running in GDB doesn't produce it, so need to use a core dump to trace it.

The debug log:
```
CHIP_DEVICE_TYPE=cpu CHIP_LOGLEVEL=debug /home/pjaaskel/src/chip-spv/build/catch/unit/streamperthread/hipStreamPerThread_MultiThread "Unit_hipStreamPerThread_MultiThread"
CHIP debug [TID 179597] [1664891123.666411586] : Reading CHIP_PLATFORM from env
CHIP debug [TID 179597] [1664891123.666472359] : Reading CHIP_DEVICE_TYPE from env
CHIP debug [TID 179597] [1664891123.666480533] : Reading CHIP_DEVICE from env
CHIP debug [TID 179597] [1664891123.666484667] : Reading CHIP_BE from env
CHIP debug [TID 179597] [1664891123.666494466] : CHIP_PLATFORM=0
CHIP debug [TID 179597] [1664891123.666498063] : CHIP_DEVICE_TYPE=cpu
CHIP debug [TID 179597] [1664891123.666501338] : CHIP_DEVICE=0
CHIP debug [TID 179597] [1664891123.666504597] : CHIP_BE=opencl
CHIP debug [TID 179597] [1664891123.666510127] : CHIPDriver Initialize
CHIP debug [TID 179597] [1664891123.666513626] : CHIPBE=OPENCL... Initializing OpenCL Backend
CHIP debug [TID 179597] [1664891123.666548393] : CHIPBackend Base Constructor
CHIP debug [TID 179597] [1664891123.755828897] : CHIPContext.add_device() 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
CHIP debug [TID 179597] [1664891123.755868295] : CHIPDevice.add_device() 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
CHIP debug [TID 179597] [1664891123.755895438] : CHIPDevice 0x555cc85d6910 added to the queue vector for backend 0x555cc71b1610 
CHIP debug [TID 179597] [1664891123.755909580] : Reading CHIP_JIT_FLAGS from env
CHIP debug [TID 179597] [1664891123.755923890] : __hipRegisterFatBinary
CHIP debug [TID 179597] [1664891123.755942087] : Bundle entry ID 0 is: 'host-x86_64-unknown-linux'

CHIP debug [TID 179597] [1664891123.755946782] : not a SPIR-V triple, ignoring

CHIP debug [TID 179597] [1664891123.755951145] : Bundle entry ID 1 is: 'hip-spirv64----generic'

CHIP debug [TID 179597] [1664891123.755979532] : Register module: 0x555cc8541d30 

CHIP debug [TID 179597] [1664891123.755983881] : CHIPBackend->register_module()
CHIP debug [TID 179597] [1664891123.756234531] : JIT compiler flags: -x spir -cl-kernel-arg-info
CHIP debug [TID 179597] [1664891123.762231677] : __hipRegisterFatBinary
CHIP debug [TID 179597] [1664891123.762259990] : Bundle entry ID 0 is: 'host-x86_64-unknown-linux'

CHIP debug [TID 179597] [1664891123.762264324] : not a SPIR-V triple, ignoring

CHIP debug [TID 179597] [1664891123.762268468] : Bundle entry ID 1 is: 'hip-spirv64----generic'

CHIP debug [TID 179597] [1664891123.762288964] : Register module: 0x555cc862ae00 

CHIP debug [TID 179597] [1664891123.762292326] : CHIPBackend->register_module()
CHIP debug [TID 179597] [1664891123.762353351] : JIT compiler flags: -x spir -cl-kernel-arg-info
Filters: Unit_hipStreamPerThread_MultiThread
CHIP debug [TID 179608] [1664891123.766131009] : Currently used memory on dev 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz: 0 M

===============================================================================
test cases: 1 | 1 passed
assertions: - none -

CHIP debug [TID 179608] [1664891123.766334156] : CHIPAllocationTracker::recordAllocation size: 131072 HOST 0x0 DEV 0x7f5848065000 TYPE 0
CHIP debug [TID 179608] [1664891123.766347569] : Currently used memory on dev 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz: 0 M

CHIP debug [TID 179608] [1664891123.766380213] : CHIPAllocationTracker::recordAllocation size: 131072 HOST 0x0 DEV 0x7f5848045000 TYPE 1
CHIP debug [TID 179609] [1664891123.766402462] : Currently used memory on dev 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz: 0 M

CHIP debug [TID 179609] [1664891123.766452770] : CHIPAllocationTracker::recordAllocation size: 131072 HOST 0x0 DEV 0x7f5848025000 TYPE 0
CHIP debug [TID 179608] [1664891123.766463648] : PerThreadDefaultQueue is null.. Creating a new queue.
CHIP debug [TID 179609] [1664891123.766468718] : Currently used memory on dev 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz: 0 M

CHIP warning [TID 179608] [1664891123.766473767] : CHIPQueueOpenCL is ignoring Priority value
CHIP debug [TID 179609] [1664891123.766490726] : CHIPAllocationTracker::recordAllocation size: 131072 HOST 0x0 DEV 0x7f5848005000 TYPE 1
CHIP debug [TID 179612] [1664891123.766521835] : Currently used memory on dev 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz: 0 M

CHIP debug [TID 179597] [1664891123.766538472] : Unregister module: 0x555cc862ae00 

CHIP debug [TID 179597] [1664891123.766553650] : CHIPBackend->unregister_module()
CHIP debug [TID 179612] [1664891123.766568969] : CHIPAllocationTracker::recordAllocation size: 131072 HOST 0x0 DEV 0x7f583c6a2000 TYPE 0
CHIP debug [TID 179612] [1664891123.766582203] : Currently used memory on dev 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz: 0 M

CHIP debug [TID 179597] [1664891123.766583902] : __hipUnRegisterFatBinary 1

CHIP debug [TID 179612] [1664891123.766597333] : CHIPAllocationTracker::recordAllocation size: 131072 HOST 0x0 DEV 0x7f583c682000 TYPE 1
CHIP debug [TID 179609] [1664891123.766609434] : PerThreadDefaultQueue is null.. Creating a new queue.
CHIP debug [TID 179597] [1664891123.766602880] : Unregister module: 0x555cc8541d30 

CHIP debug [TID 179597] [1664891123.766628018] : CHIPBackend->unregister_module()
CHIP warning [TID 179609] [1664891123.766620069] : CHIPQueueOpenCL is ignoring Priority value
CHIP debug [TID 179597] [1664891123.766636436] : __hipUnRegisterFatBinary 0

CHIP debug [TID 179597] [1664891123.766658234] : Uninitializing CHIP...
CHIP debug [TID 179597] [1664891123.766664867] : CHIPBackend::uninitialize()
CHIP debug [TID 179613] [1664891123.766667473] : Currently used memory on dev 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz: 0 M

CHIP debug [TID 179612] [1664891123.766708444] : PerThreadDefaultQueue is null.. Creating a new queue.
CHIP warning [TID 179612] [1664891123.766714707] : CHIPQueueOpenCL is ignoring Priority value
CHIP debug [TID 179613] [1664891123.766721893] : CHIPAllocationTracker::recordAllocation size: 131072 HOST 0x0 DEV 0x7f583c662000 TYPE 0
CHIP debug [TID 179613] [1664891123.766733936] : Currently used memory on dev 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz: 1 M

CHIP debug [TID 179613] [1664891123.766745645] : CHIPAllocationTracker::recordAllocation size: 131072 HOST 0x0 DEV 0x7f583c642000 TYPE 1
CHIP debug [TID 179615] [1664891123.766760333] : Currently used memory on dev 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz: 1 M

CHIP debug [TID 179615] [1664891123.766791612] : CHIPAllocationTracker::recordAllocation size: 131072 HOST 0x0 DEV 0x7f583c622000 TYPE 0
CHIP debug [TID 179615] [1664891123.766802721] : Currently used memory on dev 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz: 1 M

CHIP debug [TID 179613] [1664891123.766815355] : PerThreadDefaultQueue is null.. Creating a new queue.
CHIP warning [TID 179613] [1664891123.766822049] : CHIPQueueOpenCL is ignoring Priority value
CHIP debug [TID 179615] [1664891123.766818126] : CHIPAllocationTracker::recordAllocation size: 131072 HOST 0x0 DEV 0x7f58370e5000 TYPE 1
CHIP debug [TID 179607] [1664891123.766843381] : Currently used memory on dev 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz: 1 M

CHIP debug [TID 179607] [1664891123.766865770] : CHIPAllocationTracker::recordAllocation size: 131072 HOST 0x0 DEV 0x7f58370c5000 TYPE 0
CHIP debug [TID 179607] [1664891123.766875181] : Currently used memory on dev 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz: 1 M

CHIP debug [TID 179607] [1664891123.766885786] : CHIPAllocationTracker::recordAllocation size: 131072 HOST 0x0 DEV 0x7f58370a5000 TYPE 1
CHIP debug [TID 179615] [1664891123.766951987] : PerThreadDefaultQueue is null.. Creating a new queue.
CHIP warning [TID 179615] [1664891123.766959417] : CHIPQueueOpenCL is ignoring Priority value
CHIP debug [TID 179607] [1664891123.766977388] : PerThreadDefaultQueue is null.. Creating a new queue.
CHIP warning [TID 179607] [1664891123.766987384] : CHIPQueueOpenCL is ignoring Priority value
CHIP debug [TID 179608] [1664891123.766982482] : Setting LastEvent for 0x7f57fc001ed0 0x0 -> 0x7f57fc002a60
CHIP debug [TID 179608] [1664891123.767098176] : CHIPEventOpenCL::increaseRefCount() 0x7f57fc002a60 barrierSyncQueue refc 0->1 REASON: updateLastEvent - new event
CHIP debug [TID 179608] [1664891123.767113726] : CHIPEventOpenCL::increaseRefCount() 0x7f57fc002a60 OpenCL RefCount: 2
CHIP debug [TID 179606] [1664891123.767163060] : Currently used memory on dev 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz: 1 M

CHIP debug [TID 179606] [1664891123.767205765] : CHIPAllocationTracker::recordAllocation size: 131072 HOST 0x0 DEV 0x7f5837085000 TYPE 0
CHIP debug [TID 179606] [1664891123.767217764] : Currently used memory on dev 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz: 1 M

CHIP debug [TID 179606] [1664891123.767237273] : CHIPAllocationTracker::recordAllocation size: 131072 HOST 0x0 DEV 0x7f5836c5e000 TYPE 1
CHIP debug [TID 179612] [1664891123.767259665] : Setting LastEvent for 0x7f57f8001e60 0x0 -> 0x7f57f80029f0
hipStreamPerThread_MultiThread: /home/pjaaskel/src/chip-spv/src/backend/OpenCL/CHIPBackendOpenCL.cc:607: virtual void CHIPEventOpenCL::increaseRefCount(std::string): Assertion `status == 0' failed.
Aborted (core dumped)
```
@pvelesko any pointers?

**Comment by pvelesko** [comment]
Duplicate issue. I think I put some notes on the other tickets. Either way I spent a fair amount of time debugging and my conclusion is that we need to write a cpp reproducer for testing thread safety.

**Comment by pjaaskel** [comment]
Can you summarize these notes somewhere and explain what type of thread safety issues you suspect are there currently? What do you mean by a cpp reproducer? The test case is written in C++ as far as I understood.

**Comment by pvelesko** [comment]
Running this on the CPU causes the OpenCL event refcount assert:

```
pvelesko@iris09:~/space/CHIP-SPV/build> CHIP_BE=opencl CHIP_DEVICE_TYPE=cpu CHIP_LOGLEVEL=err ./catch/unit/streamperthread/hipStreamPerThread_MultiThread "Unit_hipStreamPerThread_MultiThread"
Filters: Unit_hipStreamPerThread_MultiThread
===============================================================================
test cases: 1 | 1 passed
assertions: - none -

CHIP error [TID 29310] [1665584662.252516550] : hipErrorNotInitialized (CL_UNKNOWN_ERROR ) in /gpfs/jlse-fs0/users/pvelesko/CHIP-SPV/src/backend/OpenCL/CHIPBackendOpenCL.cc:931:CHIPQueueOpenCL

CHIP error [TID 29307] [1665584662.252747821] : hipErrorNotInitialized (CL_UNKNOWN_ERROR ) in /gpfs/jlse-fs0/users/pvelesko/CHIP-SPV/src/backend/OpenCL/CHIPBackendOpenCL.cc:931:CHIPQueueOpenCL

CHIP error [TID 29310] [1665584662.252749531] : Caught Error: hipErrorNotInitialized
CHIP error [TID 29307] [1665584662.252818279] : Caught Error: hipErrorNotInitialized
hipStreamPerThread_MultiThread: /gpfs/jlse-fs0/users/pvelesko/CHIP-SPV/src/backend/OpenCL/CHIPBackendOpenCL.cc:609: virtual void CHIPEventOpenCL::increaseRefCount(std::string): Assertion `status == 0' failed.
Aborted
```

Running on the GPU sometimes segfaults at uninit

**Comment by pvelesko** [comment]
GPU

```
#0  0x00007ffff2e23bc4 in NEO::EmbeddedStorageRegistry::get(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const ()
   from /soft/libraries/intel-gpu-umd/f81b779-2022.06.09/driver/lib64/intel-opencl/libigdrcl.so
#1  0x00007ffff2e23c72 in NEO::EmbeddedStorage::loadImpl(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) () from /soft/libraries/intel-gpu-umd/f81b779-2022.06.09/driver/lib64/intel-opencl/libigdrcl.so
#2  0x00007ffff2e23b23 in NEO::Storage::load(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) () from /soft/libraries/intel-gpu-umd/f81b779-2022.06.09/driver/lib64/intel-opencl/libigdrcl.so
#3  0x00007ffff2e24685 in NEO::BuiltinsLib::getBuiltinResource(unsigned int, NEO::BuiltinCode::ECodeType, NEO::Device&) () from /soft/libraries/intel-gpu-umd/f81b779-2022.06.09/driver/lib64/intel-opencl/libigdrcl.so
#4  0x00007ffff2e24d6b in NEO::BuiltinsLib::getBuiltinCode(unsigned int, NEO::BuiltinCode::ECodeType, NEO::Device&) () from /soft/libraries/intel-gpu-umd/f81b779-2022.06.09/driver/lib64/intel-opencl/libigdrcl.so
#5  0x00007ffff2c2a6f3 in void NEO::BuiltinDispatchInfoBuilder::populate<char const (&) [31], NEO::MultiDeviceKernel*&, char const (&) [25], NEO::MultiDeviceKernel*&, char const (&) [35], NEO::MultiDeviceKernel*&, char const (&) [32], NEO::MultiDeviceKernel*&>(unsigned int, NEO::ConstStringRef, char const (&) [31], NEO::MultiDeviceKernel*&, char const (&) [25], NEO::MultiDeviceKernel*&, char const (&) [35], NEO::MultiDeviceKernel*&, char const (&) [32], NEO::MultiDeviceKernel*&)
    () from /soft/libraries/intel-gpu-umd/f81b779-2022.06.09/driver/lib64/intel-opencl/libigdrcl.so
#6  0x00007ffff2c1fe8b in std::call_once<NEO::BuiltInDispatchBuilderOp::getBuiltinDispatchInfoBuilder(unsigned int, NEO::ClDevice&)::{lambda()#1}>(std::once_flag&, NEO::BuiltInDispatchBuilderOp::getBuiltinDispatchInfoBuilder(unsigned int, NEO::ClDevice&)::{lambda()#1}&&)::{lambda()#2}::_FUN() () from /soft/libraries/intel-gpu-umd/f81b779-2022.06.09/driver/lib64/intel-opencl/libigdrcl.so
#7  0x00007ffff68145a7 in __pthread_once_slow () from /lib64/libpthread.so.0
#8  0x00007ffff2c1ea0c in NEO::BuiltInDispatchBuilderOp::getBuiltinDispatchInfoBuilder(unsigned int, NEO::ClDevice&) () from /soft/libraries/intel-gpu-umd/f81b779-2022.06.09/driver/lib64/intel-opencl/libigdrcl.so
#9  0x00007ffff2a072a0 in NEO::CommandQueueHw<NEO::SKLFamily>::enqueueSVMMemcpy(unsigned int, void*, void const*, unsigned long, unsigned int, _cl_event* const*, _cl_event**) ()
   from /soft/libraries/intel-gpu-umd/f81b779-2022.06.09/driver/lib64/intel-opencl/libigdrcl.so
#10 0x00007ffff28ce0f8 in clEnqueueSVMMemcpy () from /soft/libraries/intel-gpu-umd/f81b779-2022.06.09/driver/lib64/intel-opencl/libigdrcl.so
#11 0x00007ffff7ab8778 in CHIPQueueOpenCL::memCopyAsyncImpl (this=0x7fffd4043a10, Dst=0x7fffd4023000, Src=0x7fffd4001000, Size=131072) at /home/pvelesko/space/CHIP-SPV/src/backend/OpenCL/CHIPBackendOpenCL.cc:955
#12 0x00007ffff7a1a687 in CHIPQueue::memCopyAsync (this=0x7fffd4043a10, Dst=0x7fffd4023000, Src=0x7fffd4001000, Size=131072) at /home/pvelesko/space/CHIP-SPV/src/CHIPBackend.cc:1490
#13 0x00007ffff7a606f1 in hipMemcpyAsync (Dst=0x7fffd4023000, Src=0x7fffd4001000, SizeBytes=131072, Kind=hipMemcpyHostToDevice, Stream=0x7fffd4043a10) at /home/pvelesko/space/CHIP-SPV/src/CHIPBindings.cc:2086
#14 0x00005555555a4585 in Copy_to_device () at /gpfs/jlse-fs0/users/pvelesko/CHIP-SPV/HIP/tests/catch/unit/streamperthread/hipStreamPerThread_MultiThread.cc:35
#15 0x00005555555a53e2 in std::__invoke_impl<void, void (*)()> (__f=@0x5555566f9048: 0x5555555a44f0 <Copy_to_device()>) at /usr/lib64/gcc/x86_64-suse-linux/7/../../../../include/c++/7/bits/invoke.h:60
#16 0x00005555555a53a5 in std::__invoke<void (*)()> (__fn=@0x5555566f9048: 0x5555555a44f0 <Copy_to_device()>) at /usr/lib64/gcc/x86_64-suse-linux/7/../../../../include/c++/7/bits/invoke.h:95
#17 0x00005555555a537d in std::thread::_Invoker<std::tuple<void (*)()> >::_M_invoke<0ul> (this=0x5555566f9048) at /usr/lib64/gcc/x86_64-suse-linux/7/../../../../include/c++/7/thread:234
#18 0x00005555555a5355 in std::thread::_Invoker<std::tuple<void (*)()> >::operator() (this=0x5555566f9048) at /usr/lib64/gcc/x86_64-suse-linux/7/../../../../include/c++/7/thread:243
#19 0x00005555555a5279 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)()> > >::_M_run (this=0x5555566f9040) at /usr/lib64/gcc/x86_64-suse-linux/7/../../../../include/c++/7/thread:186
#20 0x00007ffff705d4f0 in std::execute_native_thread_routine (__p=0x5555566f9040) at /dev/shm/spack-stage-servesh/spack-stage-gcc-10.2.0-yudlyezca7twgd5o3wkkraur7wdbngdn/spack-src/libstdc++-v3/src/c++11/thread.cc:80
#21 0x00007ffff680b6ea in start_thread () from /lib64/libpthread.so.0
#22 0x00007ffff6523a8f in clone () from /lib64/libc.so.6
```

**Comment by pvelesko** [comment]
Inside `SVMemoryRegion::allocate()`, I added a check to get the context reference count. When the allocation takes place, the context reference count query returns CL_SUCCESS but the count is 32525 

@linehill it would be great overall to migrate away from C++ OpenCL bindings completely that was the source of reference counting issues with events previously.

**Comment by linehill** [comment]
> @linehill maybe you can take a look since you wrote this.

It's originally written by @franz but I have done minor changes to it. Perhaps, he could take a look.

**Comment by pjaaskel** [comment]
@pvelesko do you mean there is a bug in the C++ OpenCL bindings? We should report if so.

**Comment by pjaaskel** [comment]
Any news related to this? It's still failing for both CPU and GPU. Are there known problems with refcounting? Or is this the testcase discussed on Monday which is likely a testcase problem due to unclean thread exit?

**Comment by pjaaskel** [comment]
Unit_hipStreamPerThread_DeviceReset_1 might crash due to the same issue.

**Comment by pvelesko** [comment]
Working on it.. #209  It's gonna end up being a huge PR
